DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs

```diff
@@ -0,0 +1,188 @@
+
+        public NonBlittableArrayMarshaller(IMarshallingGenerator elementMarshaller, ExpressionSyntax numElementsExpr)
+        {
+            _elementMarshaller = elementMarshaller;
+            _numElementsExpr = numElementsExpr;
+        }
+
+        private ITypeSymbol GetElementTypeSymbol(TypePositionInfo info)
+        {
+            return ((IArrayTypeSymbol)info.ManagedType).ElementType;
+        }
+
+        private TypeSyntax GetNativeElementTypeSyntax(TypePositionInfo info)
+        {
+            return _elementMarshaller.AsNativeType(TypePositionInfo.CreateForType(GetElementTypeSymbol(info), NoMarshallingInfo.Instance));
+        }
+
+        public override TypeSyntax AsNativeType(TypePositionInfo info)
+        {
+            return PointerType(GetNativeElementTypeSyntax(info));
+        }
+
+        public override ParameterSyntax AsParameter(TypePositionInfo info)
+        {
+            var type = info.IsByRef
+                ? PointerType(AsNativeType(info))
+                : AsNativeType(info);
+            return Parameter(Identifier(info.InstanceIdentifier))
+                .WithType(type);
+        }
+
+        public override ArgumentSyntax AsArgument(TypePositionInfo info, StubCodeContext context)
+        {
+            return info.IsByRef
+                ? Argument(
+                    PrefixUnaryExpression(
+                        SyntaxKind.AddressOfExpression,
+                        IdentifierName(context.GetIdentifiers(info).native)))
+                : Argument(IdentifierName(context.GetIdentifiers(info).native));
+        }
+
+        public override IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeContext context)
+        {
+            var (managedIdentifer, nativeIdentifier) = context.GetIdentifiers(info);
+
+            switch (context.CurrentStage)
+            {
+                case StubCodeContext.Stage.Setup:
+                    yield return LocalDeclarationStatement(
+                        VariableDeclaration(
+                            AsNativeType(info),
+                            SingletonSeparatedList(VariableDeclarator(nativeIdentifier))));
+                    break;
+                case StubCodeContext.Stage.Marshal:
+                    if (info.RefKind != RefKind.Out)
+                    {
+                        foreach (var statement in GenerateConditionalAllocationSyntax(
+                            info,
+                            context,
+                            StackAllocBytesThreshold,
+                            checkForNull: true))
+                        {
+                            yield return statement;
+                        }
+
+                        string indexerIdentifier = "i";
+                        var arraySubContext = new ArrayMarshallingCodeContext(context.CurrentStage, indexerIdentifier, context);
+                        yield return MarshallerHelpers.GetForLoop(managedIdentifer, indexerIdentifier)
```

Comment:
Null check before doing the loop too, since it does `<identifier>.Length`? Also in Cleanup.

Reply:
We should also have tests with null array parameter in / out.